src/main/java/io/divolte/server/filesinks/gcs/GoogleCloudStorageFileManager.java

```diff
@@ -0,0 +1,406 @@
+        final HttpURLConnection connection = setupUrlConnection(POST, url, true, additionalHeaders);
+
+        final OutputStream os = connection.getOutputStream();
+        writer.write(os);
+        os.flush();
+        os.close();
+
+        return parseResponse(resultType, connection);
+    }
+
+    private interface BodyWriter {
+        void write(final OutputStream stream) throws IOException;
+    }
+
+    private static <T> T googleGet(final URL url, final Class<T> resultType) throws IOException {
+        return parseResponse(resultType, setupUrlConnection(GET, url, false, Collections.emptyMap()));
+    }
+
+    @SuppressWarnings("PMD.EmptyWhileStmt")
+    private static void googleDelete(final URL url) throws IOException {
+        final HttpURLConnection connection = setupUrlConnection(DELETE, url, false, Collections.emptyMap());
+        final int responseCode = connection.getResponseCode();
+        if (responseCode < 200 || responseCode > 299) {
+            final InputStream stream = connection.getErrorStream();
+            // Read the error response as String; GCS sometimes sends a JSON error response
+            // and sometimes text/plain (e.g. "Not found.")
+            final String response = CharStreams.toString(new InputStreamReader(stream, URL_ENCODING));
+            stream.close();
+            logger.error("Received unexpected response from Google Cloud Storage. Response status code: {}. Response body: {}", responseCode, response);
+            throw new IOException("Received unexpected response from Google Cloud Storage.");
+        } else {
+            final InputStream stream = connection.getInputStream();
+            /*
+             * As per the docs, Google sends a empty response with a 200. We need to get and
+             * drain the stream for the HTTP client to consider the connection for reuse.
+             */
+            while (stream.read() != -1) {}
+        }
+    }
+
+    private static <T> T parseResponse(final Class<T> resultType, final HttpURLConnection connection) throws IOException, JsonParseException, JsonMappingException {
+        final int responseCode = connection.getResponseCode();
+        // Note: the docs are specific about closing the streams after reading in order
+        // to trigger proper Keep-Alive usage
+        if (responseCode < 200 || responseCode > 299) {
+            final InputStream stream = connection.getErrorStream();
+            // Read the error response as String; GCS sometimes sends a JSON error response
+            // and sometimes text/plain (e.g. "Not found.")
+            final String response = CharStreams.toString(new InputStreamReader(stream, URL_ENCODING));
+            stream.close();
+            logger.error("Received unexpected response from Google Cloud Storage. Response status code: {}. Response body: {}", responseCode, response);
+            throw new IOException("Received unexpected response from Google Cloud Storage.");
+        } else {
```

Comment:
The else branch wasn't necessary in the first place, as the first branch of the if-statement throws `IOException`. I extracted the check and throwing `IOException` to a method named `throwIOExceptionOnErrorResponse` and call that as a guard before continuing on the happy flow.